### DISCUSSION TITLE
Create a binary with statically linked dependencies

### DISCUSSION THREAD

**Discussion by hovsater** [universal-ctags/ctags#2977] [Q&A] [answered]
Hey! 👋 I'm looking into updating the ctags binary in https://github.com/textmate/ctags.tmbundle to the latest version. The [TextMate](https://macromates.com) bundle seem to have a statically linked binary of `ctags`.

```console
kevinsjoberg@Kevins-MacBook-Pro ~/c/ctags.tmbundle (master)> otool -L Support/bin/ctags-objc 
Support/bin/ctags-objc:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 159.1.0)
```

I'm fairly inexperienced when it comes to C and its tooling, so I'm not 100% sure how I would go about building a statically linked binary. It only needs to run on macOS since TextMate is a macOS application.

I found a flag, `--enable-static`, which I assume I could use in some way to achieve what I'm after. Any pointers would be much appreciated!

**Comment by leleliu008** [comment, marked as answer]
you can not use `--enable-static` flags on `macOS`, because `macOS` not provide `/usr/lib/libSystem.B.a`

**Reply by hovsater** [reply]
@leleliu008 thanks for explaining. Looking at the output from `otool` it looks like `/usr/lib/libSystem.B.a` is still dynamically linked, while the other dependencies are not. Is there a way to achieve that on macOS?

Given it's going to be run on macOS only, it's okay that `/usr/lib/libSystem.B.a` is dynamically linked while the other dependencies are not.

**Reply by leleliu008** [reply]
There are many solutions. I have been working on cmake build system. this issue can be easily resolved if using cmake build system. But you need wait one day, I will finish this work tomorrow.

**Reply by hovsater** [reply]
Thanks. Appreciate the help! 🙂

**Reply by leleliu008** [reply]
**step1: install dependency libraries and tools**
```bash
brew install jansson libxml2 libyaml docutils
```
**step2: clone universal-ctags**
```bash
git clone https://github.com/universal-ctags/ctags
```
**step4: change to ctags dir**
```bash
cd ctags
```
**step5: clone cmakew**
```bash
git clone https://github.com/leleliu008/cmakew
```
**step6: copy cmakew/\* to ctags dir**
```bash
cp -r cmakew/* .
```
**step7: apply patches**
```bash
patch -p1  < patch/*.patch
```
**step8: config**
```bash
./cmakew config -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=./output
```
**step9: build and install**
```bash
./cmakew install
```

**Reply by hovsater** [reply]
@leleliu008 thank you! 🙏 Seems to be working just fine!

**Comment by leleliu008** [comment]
`--enable-static` can be used on any other platform except `macOS `. using this flags is not a good idea, because this flags cause `libc.a` linked into `ctags` binary. It may cause problems when you distribute to other os.

**Reply by hovsater** [reply]
I'm not going to distribute this on any other platforms. I need a binary that can be run on any **macOS** machine without relying on the dependencies to be available at a certain path. I want the binary to work without having to tell the consumer to install `jansson` etc to make it work.

**Comment by psprint** [comment]
Anyway, a binary via GitHub releases would be cool, allowing installation via zinit zsh plugin manager, like dstask, fd, bat, fzf and others.

**Reply by masatake** [reply]
Like ctags-win32 repo, you can run a repository + github actions to build and distribute binary files.

**Reply by leleliu008** [reply]
this may helps: https://github.com/leleliu008/uctags-nightly-build

**Reply by masatake** [reply]
@leleliu008 great! If you are o.k., you can run the repository at universal-ctags organization.

**Reply by masatake** [reply]
or you can put a link to your repository on our web site or README.

**Reply by leleliu008** [reply]
> @leleliu008 great! If you are o.k., you can run the repository at universal-ctags organization.

I'm glad to do that. 

notes:
1. There is no prebuild binaries for `macos-arm64`, becuses there is currently no GitHub Actions images for  `macos-arm64`.
2. There is no prebuild binaries for `linux-arm64`, becuses GitHub Actions images for  `macos-arm64` is slow, I will try to add this feature if the build time is acceptable.

**Reply by leleliu008** [reply]
prebuild binaries for `macos-arm64` will be supported later via cross-compling, currently, [ppkg](https://github.com/leleliu008/ppkg) which is heavily relied on by https://github.com/leleliu008/uctags-nightly-build does not support cross-compling yet.

**Reply by leleliu008** [reply]
@masatake https://github.com/leleliu008/uctags-nightly-build is ready, if you have any advise please let me known.

**Reply by masatake** [reply]
@leleliu008 I think you can create a repository under https://github.com/universal-ctags. You can run the repository like https://github.com/universal-ctags/ctags-win32.
Then you can update https://github.com/universal-ctags/ctags/blob/master/README.md .

**Reply by leleliu008** [reply]
@masatake https://github.com/leleliu008/uctags-nightly-build can be transformed to https://github.com/universal-ctags

Reference: https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository

**Reply by leleliu008** [reply]
@masatake I transformed https://github.com/leleliu008/uctags-nightly-build to https://github.com/universal-ctags/uctags-nightly-build yesterday, but it has limitation, so I create a new repository https://github.com/universal-ctags/ctags-nightly-build today, so https://github.com/universal-ctags/uctags-nightly-build can be deleted, I have insufficient permission to delete repository it, I think you can delete it.

**Reply by masatake** [reply]
Done. Thank you.
